util/rateLimit.go

```diff
@@ -0,0 +1,501 @@
+				if consumedByClient[e.c] == nil {
+					ts := []time.Time{}
+					consumedByClient[e.c] = &ts
+				}
+				*(consumedByClient[e.c]) = append(*(consumedByClient[e.c]), e.t)
+			// served events -- the capacity has been totally served
+			case e := <-cm.served:
+				serves = append(serves, e.t)
+			// check for context Done, and flag the thread for shutdown
+			case <-cm.ctx.Done():
+				exit = true
+			}
+
+		}
+		// only recalculate the service rate every N ticks, because all lists must be pruned, which can be expensive
+		// also calculate if the routine is going to exit
+		tick = (tick + 1) % cm.targetRateRefreshTicks
+		if tick == 0 || exit {
+			cutoff := time.Now().Add(-1 * cm.window)
+			prune(&serves, cutoff)
+			for c := range consumedByClient {
+				if prune(consumedByClient[c], cutoff) == 0 {
+					delete(consumedByClient, c)
+				}
+			}
+			targetRate = 0
+			// targetRate is the average service rate per client per second
+			if len(consumedByClient) > 0 {
+				serviceRate := float64(len(serves)) / float64(cm.window/time.Second)
+				targetRate = serviceRate / float64(len(consumedByClient))
+			}
+		}
+		if exit {
+			cm.setTargetRate(targetRate)
+			cm.setConsumedByClient(consumedByClient)
+			cm.setServes(serves)
+			cm.runLock.Lock()
+			defer cm.runLock.Unlock()
+			cm.running = false
+			cm.wg.Done()
+			return
+		}
+	}
+}
+
+func (cm *redCongestionManager) Stop() {
+	cm.ctxCancel()
+	cm.wg.Wait()
+}
+
+func (cm *redCongestionManager) setTargetRate(tr float64) {
+	cm.targetRate = tr
+}
+
+func (cm *redCongestionManager) setConsumedByClient(cbc map[ErlClient]*[]time.Time) {
+	cm.consumedByClient = cbc
+}
+
+func (cm *redCongestionManager) setServes(ts []time.Time) {
+	cm.serves = ts
+}
+
+func (cm *redCongestionManager) arrivalRateFor(arrivals *[]time.Time) float64 {
+	clientArrivalRate := float64(0)
+	if arrivals != nil {
+		clientArrivalRate = float64(len(*arrivals)) / float64(cm.window/time.Second)
+	}
+	return clientArrivalRate
+}
+
+func (cm *redCongestionManager) shouldDrop(targetRate float64, c ErlClient, arrivals *[]time.Time) bool {
+	// clients who have "never" been seen do not get dropped
+	clientArrivalRate := cm.arrivalRateFor(arrivals)
+	if clientArrivalRate == 0 {
+		return false
+	}
+	// if targetRate is 0, it means we haven't had any activity to calculate (or there is not enough data)
+	// it should not drop in this case
+	if targetRate == 0 {
+		return false
+	}
+	// A random float is selected, and the arrival rate of the given client is
+	// turned to a ratio against targetRate. the congestion manager recommends to drop activity
+	// proportional to its overuse above the targetRate
+	r := rand.Float64()
+	return (math.Pow(clientArrivalRate, cm.exp) / math.Pow(targetRate, cm.exp)) > r
+}
+
+func prune(ts *[]time.Time, cutoff time.Time) int {
+	if ts == nil {
+		return 0
+	}
+	// find the first inserted timestamp *after* the cutoff, and cut everything behind it off
+	for i, t := range *ts {
```

Comment:
optimization ideas: 
1) check the last ts element if it is before the cutoff: then can cut all of them without iteration through the array
2) for the arrays longer than say 50 elements use bin search

Reply:
Took both of these, great ideas thanks!